code/obj/machinery/dialysis.dm

```diff
@@ -0,0 +1,172 @@
+		if (src.reagents.is_full())
+			src.reagents.clear_reagents()
+			src.audible_message("<span class='game say'><span class='name'>[src]</span> beeps, \"Blood buffer has reached maximum capacity. Purging internal reservoir.\"")
+
+		transfer_blood(src.patient, src, src.draw_amount)
+
+		// Re-implemented here due to all the got dang boutputs.
+		var/list/whitelist_buffer = src.whitelist + src.patient_blood_id
+		for (var/reagent_id in src.reagents.reagent_list)
+			if ((!src.hacked && !(reagent_id in whitelist_buffer)) || (src.hacked && (reagent_id in whitelist_buffer)))
+				src.reagents.del_reagent(reagent_id)
+
+		src.output_blood_colour = src.reagents.total_volume ? src.reagents.get_average_color().to_rgba() : null
+
+		// Infuse blood back in if possible. Don't wanna stuff too much blood back in.
+		// The blood that's not actually in the bloodstream yet, know what I mean?
+		var/non_blood_volume_blood = src.patient.reagents.reagent_list["blood"]?.volume || 0
+		if (!src.patient.reagents.is_full() && (src.patient.blood_volume + non_blood_volume_blood) <= 500)
+			src.reagents.trans_to(src.patient, src.draw_amount)
+			src.patient.reagents.reaction(src.patient, INGEST, src.draw_amount)
+
+		src.UpdateIcon()
+
+	update_icon(...)
+		..()
+		if (src.patient)
+			src.UpdateOverlays(image(src.icon, "cannulae"), "tubing")
+
+			var/image/blood_out = image(src.icon, "tubing-good[src.output_blood_colour ? "" : "-empty"]")
+			if (src.output_blood_colour)
+				blood_out.color = src.output_blood_colour
+			src.UpdateOverlays(blood_out, "blood_out")
+
+			var/image/blood_in = image(src.icon, "tubing-bad[(src.patient.blood_volume || src.patient.reagents.total_volume) ? "" : "-empty"]")
+			if (src.patient.reagents.total_volume)
+				blood_in.color = src.patient.reagents.get_average_color().to_rgba()
+			else
+				blood_in.color = src.patient.blood_color
+			src.UpdateOverlays(blood_in, "blood_in")
+		else
+			src.UpdateOverlays(image(src.icon, "pump-off"), "pump")
+			src.UpdateOverlays(image(src.icon, "screen-off"), "screen")
+			src.UpdateOverlays(image(src.icon, "tubing"), "tubing")
+			src.ClearSpecificOverlays("blood_out")
+			src.ClearSpecificOverlays("blood_in")
+
+	proc/cannulate(mob/living/carbon/new_patient, mob/user)
+		src.patient = new_patient
+		new_patient.tri_message(user,\
+			"<span class='notice'><b>[user]</b> inserts [src]'s cannulae into [new_patient == user ? "[his_or_her(new_patient)]" : "[new_patient]'s"] arm.</span>",\
+			"<span class='notice'>[new_patient == user ? "You insert" : "<b>[user]</b> inserts"] [src]'s cannulae into your arm.</span>",\
+			"<span class='notice'>You insert [src]'s cannulae into [new_patient == user ? "your" : "[new_patient]'s"] arm.</span>")
+		logTheThing(LOG_COMBAT, user, "connects a dialysis machine [log_reagents(src)] to [constructTarget(new_patient,"combat")] at [log_loc(user)].")
+		src.start_dialysis()
+
+	proc/start_dialysis()
```

Comment:
I think it'd be a tiny bit cleaner if this proc got a `new_patient` argument and set `src.patient` to that. Also I don't see any handling of what happens when old patient is overwritten with a new patient. Shouldn't that at least have its own message or something?

Reply:
Patients can't (or shouldn't) be overridden without removing the current patients first, with how this has been implemented.